File: code/frontend/src/app/services/recipe.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Globals } from '../global';
import { Observable } from 'rxjs';
import { LightRecipeDto, Pageable, PlanedRecipeCreationDto, PlanedRecipeDto, RecipeCategoryDto, RecipeCreationDto, RecipeDto, RecipeIngredientListDto, RecipeRatingDto, RecipeStepsDto } from '../dtos';
import { RecipeFilterDto } from '../dtos/recipe-filter-dto';

@Injectable({
  providedIn: 'root',
})
export class RecipeService {
  private recipeBaseUri: string = this.globals.backendUri + '/recipe';

  constructor(private httpClient: HttpClient, private globals: Globals) {}

  getAllRecipeCategories(): Observable<RecipeCategoryDto[]> {
    return this.httpClient.get<RecipeCategoryDto[]>(`${this.recipeBaseUri}/categories`);
  }

  getAllRecipeTags(): Observable<string[]> {
    return this.httpClient.get<string[]>(`${this.recipeBaseUri}/tags`);
  }

  create(recipe: RecipeCreationDto): Observable<RecipeDto> {
    return this.httpClient.post<RecipeDto>(`${this.recipeBaseUri}`, recipe);
  }

  get(id: number): Observable<RecipeDto> {
    return this.httpClient.get<RecipeDto>(`${this.recipeBaseUri}/${id}`);
  }

  edit(recipe: RecipeDto, id: number): Observable<RecipeDto> {
    return this.httpClient.put<RecipeDto>(`${this.recipeBaseUri}/${id}`, recipe);
  }

  getAll(
    page_: number,
    pageSize_: number,
  ): Observable<Pageable<LightRecipeDto>> {
    return this.httpClient.get<Pageable<LightRecipeDto>>(`${this.recipeBaseUri}/all`,
      { params: { page: page_, pageSize: pageSize_ } });
  }

  delete(id: number): Observable<any> {
    return this.httpClient.delete<any>(`${this.recipeBaseUri}/${id}`);
  }

  editRecipeIngredientList(recipeIngredientList: RecipeIngredientListDto): Observable<any> {
    return this.httpClient.post<any>(`${this.recipeBaseUri}/ingredient-list`, recipeIngredientList);
  }

  getRecipeIngredientList(id: number): Observable<RecipeIngredientListDto> {
    return this.httpClient.get<RecipeIngredientListDto>(`${this.recipeBaseUri}/${id}/ingredient-list`);
  }

  updateSteps(stepsDto: RecipeStepsDto): Observable<any> {
    return this.httpClient.post<any>(`${this.recipeBaseUri}/steps`, stepsDto);
  }

  getSteps(recipeId: number): Observable<RecipeStepsDto> {
    return this.httpClient.get<RecipeStepsDto>(`${this.recipeBaseUri}/${recipeId}/steps`);
  }

  saveRating(rating: RecipeRatingDto): Observable<any> {
    return this.httpClient.post<any>(`${this.recipeBaseUri}/rating`, rating);
  }

  getRatings(
    page: number,
    pageSize: number,
    recipeId: number
  ): Observable<Pageable<RecipeRatingDto>> {
    return this.httpClient.get<Pageable<RecipeRatingDto>>(`${this.recipeBaseUri}/${recipeId}/ratings`,
      { params: { page: page, pageSize: pageSize } });
  }

  getAllFiltered(
    page_: number,
    pageSize_: number,
    filterDto: RecipeFilterDto
  ) : Observable<Pageable<LightRecipeDto>> {
    return this.httpClient.post<Pageable<LightRecipeDto>>(`${this.recipeBaseUri}/filter`, filterDto, { params: { page: page_, pageSize: pageSize_ } });
  }

  cookRecipe(
    recipeId: number,
    ingredientList: RecipeIngredientListDto
  ) : Observable<any> {
    return this.httpClient.post<any>(`${this.recipeBaseUri}/cook/${recipeId}`, ingredientList);
  }

  planRecipe(
    recipeId: number,
    planedRecipeCreationDto: PlanedRecipeCreationDto
  ): Observable<PlanedRecipeDto> {
    return this.httpClient.post<PlanedRecipeDto>(`${this.recipeBaseUri}/plan/${recipeId}`, planedRecipeCreationDto);
  }

  getPlanedRecipes(
    year: number,
    month: number,
  ): Observable<Map<number, PlanedRecipeDto[]>> {
    return this.httpClient.get<Map<number, PlanedRecipeDto[]>>(`${this.recipeBaseUri}/planned?year=${year}&month=${month}`);
  }
}

File: code/backend/src/main/java/project/backend/service/RecipeService.java
package project.backend.service;

import project.backend.dto.LightRecipeDto;
import project.backend.dto.PageableDto;
import project.backend.dto.PlanedRecipeCreationDto;
import project.backend.dto.PlanedRecipeDto;
import project.backend.dto.RecipeCategoryDto;
import project.backend.dto.RecipeCreationDto;
import project.backend.dto.RecipeDto;
import project.backend.dto.RecipeFilterDto;
import project.backend.dto.RecipeIngredientListDto;
import project.backend.dto.RecipeRatingDto;
import project.backend.dto.RecipeStepsDto;
import project.backend.exception.ValidationException;
import project.backend.exception.NotFoundException;
import project.backend.exception.ConflictException;

import java.util.List;
import java.util.Map;

public interface RecipeService {

    /**
     * Returns all recipe categories, that have been created so far.
     *
     * @return List of recipe categories
     */
    List<RecipeCategoryDto> getRecipeCategories();

    /**
     * Returns all recipe tags, that have been created so far.
     *
     * @return List of recipe tags
     */
    List<String> getRecipeTags();

    /**
     * Creates a new recipe in the persistent storage.
     * The content of the DTO is validated before the recipe is created.
     *
     * @param recipeDto The recipe to be created
     * @throws ValidationException If the DTO is not valid
     * @return The created recipe
     */
    RecipeDto createRecipe(RecipeCreationDto recipeDto);

    /**
     * Edits an existing recipe in the persistent storage.
     * The content of the DTO is validated before the recipe is edited.
     *
     * @param recipeDto The recipe to be edited with the new information
     * @param id The id of the recipe to be edited
     * @throws ValidationException If the DTO is not valid
     * @return The edited recipe
     */
    RecipeDto editRecipe(RecipeDto recipeDto, Long id);

    /**
     * Returns a specific recipe.
     *
     * @param id The id of the recipe to be returned
     * @throws NotFoundException If the given ID does not exist
     * @return The recipe with the given id
     */
    RecipeDto getRecipe(Long id);

    /**
     * Returns all recipes, that have been created so far, using pagination.
     * The page number and the size of the page are specified by the parameters.
     *
     * @param page The page number
     * @param pageSize The size of the page
     * @return The recipes of the given page
     */
    PageableDto<LightRecipeDto> getRecipes(int page, int pageSize);

    /**
     * Deletes a recipe from the persistent storage.
     *
     * @param id The id of the recipe to be deleted
     * @throws NotFoundException If the given ID does not exist
     */
    void deleteRecipe(Long id);

    /**
     * Updates the ingredient list of a recipe.
     * The content of the DTO is validated before the ingredient list is updated.
     *
     * @param recipeIngredientListDto The new ingredient list
     */
    void updateIngredientList(RecipeIngredientListDto recipeIngredientListDto);

    /**
     * Returns the ingredient list of a recipe.
     *
     * @param recipeId The id of the recipe
     * @throws NotFoundException If the given ID does not exist
     * @return The ingredient list of the recipe
     */
    RecipeIngredientListDto getIngredientList(Long recipeId);

    /**
     * Updates the steps of a recipe.
     * The content of the DTO is validated before the steps are updated.
     *
     * @throws NotFoundException If the ID of the recipe does not exist
     * @param recipeStepsDto The new steps
     */
    void updateSteps(RecipeStepsDto recipeStepsDto);

    /**
     * Returns the steps of a recipe.
     *
     * @throws NotFoundException If the ID of the recipe does not exist
     * @param recipeId The id of the recipe
     * @return The steps of the recipe
     */
    RecipeStepsDto getSteps(Long recipeId);

    /**
     * Adds a rating to a recipe.
     * The content of the DTO is validated before the rating is added.
     *
     * @param recipeRatingDto The rating to be added
     * @throws NotFoundException If the ID of the recipe does not exist
     */
    void addRating(RecipeRatingDto recipeRatingDto);

    /**
     * Returns the ratings of a recipe in a pageable format.
     *
     * @param page The page number
     * @param pageSize The size of the page
     * @param recipeId The id of the recipe
     *
     * @return The ratings of the recipe
     */
    PageableDto<RecipeRatingDto> getRatings(int page, int pageSize, Long recipeId);

    /**
     * Returns all recipes, that match the filter criteria, using pagination.
     *
     * @param page The page number
     * @param pageSize The size of the page
     * @param recipeFilterDto The filter criteria
     *
     * @return The recipes of the given page, that match the filter criteria, using pagination
     */
    PageableDto<LightRecipeDto> getFilteredRecipes(int page, int pageSize, RecipeFilterDto recipeFilterDto);

    /**
     * Remove ingredients from inventory, as if the recipe was cooked.
     * The content of the DTO is validated before the ingredients are removed.
     *
     * @param recipeId The id of the recipe
     * @param recipeIngredientListDto The ingredient list of the recipe
     *
     * @throws NotFoundException If the ID of the recipe does not exist
     * @throws ValidationException If the DTO is not valid
     * @throws ConflictException If the needed ingredient is not in inventory
     */
    void cookRecipe(Long recipeId, RecipeIngredientListDto recipeIngredientListDto);

    /**
     * Plans a recipe for a specific date and meal.
     * The content of the DTO is validated before the recipe is planned.
     *
     * @param planedRecipeCreationDto The details of the planned recipe
     * @return The planned recipe
     */
    PlanedRecipeDto planNewRecipe(PlanedRecipeCreationDto planedRecipeCreationDto);

    /**
     * Returns all planned recipes of a specific month.
     *
     * @param year the year of the month
     * @param month the month to get the planned recipes of
     *
     * @return the planned recipes of the given month
     */
    Map<Integer, List<PlanedRecipeDto>> getPlannedRecipes(int year, int month);
}
